src/mca.jl

```diff
@@ -0,0 +1,354 @@
+
+    # The underlying corresponence analysis
+    C::CA{T}
+
+    # Variable names
+    vnames::Vector{String}
+
+    # Map values to integer positions
+    rd::Vector{Dict}
+
+    # Map integer positions to values
+    dr::Vector{Dict}
+
+    # Split the variable scores into separate arrays for
+    # each variable.
+    Gv::Vector{Matrix{T}}
+
+    # Number of nominal variables
+    K::Int
+
+    # Total number of categories in all variables
+    J::Int
+
+    # Eigenvalues
+    unadjusted_eigs::Vector{Float64}
+    benzecri_eigs::Vector{Float64}
+    greenacre_eigs::Vector{Float64}
+end
+
+objectscores(mca::MCA) = mca.C.F
+variablescores(mca::MCA) = mca.Gv
+
+# Split the variable scores to a separate array for each
+# variable.
+function xsplit(G, rd)
+    K = [length(di) for di in rd]
+    Js = cumsum(K)
+    Js = vcat(1, 1 .+ Js)
+    Gv = Vector{Matrix{eltype(G)}}()
+    for j in eachindex(K)
+        g = G[Js[j]:Js[j+1]-1, :]
+        push!(Gv, g)
+    end
+    return Gv
+end
+
+# Calculate the eigenvalues with different debiasings.
+function get_eigs(I, K, J)
+    ben = zeros(length(I))
+    gra = zeros(length(I))
+    Ki = 1 / K
+    f = K / (K - 1)
+    for i in eachindex(I)
+        if I[i] > Ki
+            ben[i] = (f * (I[i] - Ki))^2
+        end
+    end
+
+    unadjusted = I ./ sum(I)
+    gt = f * (sum(abs2, I) - (J - K) / K^2)
+
+    return unadjusted, ben ./ sum(ben), ben ./ gt
+end
+
+# constructor
+
+function MCA(Z; vnames = [])
+
+    if length(vnames) == 0
+        vnames = ["v$(j)" for j = 1:size(Z, 2)]
+    end
+
+    # Get the indicator matrix
+    X, rd, dr = make_indicators(Z)
+
+    # Create the underlying correspondence analysis value
+    C = CA(X)
+
+    # Number of nominal variables
+    K = size(Z, 2)
+
+    # Total number of categories in all variables
+    J = size(X, 2)
+
+    return MCA(C, vnames, rd, dr, Matrix{Float64}[], K, J, zeros(0), zeros(0), zeros(0))
+end
+
+"""
+    mca
+
+Fit a multiple correspondence analysis using the columns of `Z` as the
+variables.  The first `d` components are retained.  If `Z` is a
+dataframe then the column names are used as variable names, otherwise
+variable names may be provided as `vnames`.
+"""
+function mca(Z, d::Int; vnames = [])
+    m = MCA(Z; vnames)
+    fit!(m, d)
+    return m
+end
+
+function fit(::Type{MCA}, Z::AbstractMatrix, d::Int; vnames = [])
+    return mca(Z, d; vnames)
+end
+
+function fit!(mca::MCA, d::Int)
+
+    fit!(mca.C, d)
+
+    # Split the variable scores into separate arrays for each variable.
+    mca.Gv = xsplit(mca.C.G, mca.rd)
+
+    una, ben, gra = get_eigs(mca.C.I, mca.J, mca.K)
+
+    mca.unadjusted_eigs = una
+    mca.benzecri_eigs = ben
+    mca.greenacre_eigs = gra
+
+    return mca
+end
+
+# Create an indicator matrix corresponding to the distinct
+# values in z.  Also returns dictionaries mapping the unique
+# values to column offsets, and mapping the column offsets
+# to the unique values.
+function make_single_indicator(z)
```

Comment:
What is the type for `z` parameter? You can use parametrized type to avoid runtime type evaluation, see below `eltype` call.